sumcheck/src/util.rs

```diff
@@ -140,73 +136,154 @@ fn inner_extrapolate<F: Field, const IS_PARALLEL: bool>(
         * sum_inv
 }
 
-/// Interpolate a uni-variate degree-`p_i.len()-1` polynomial and evaluate this
-/// polynomial at `eval_at`:
-///
-///   \sum_{i=0}^len p_i * (\prod_{j!=i} (eval_at - j)/(i-j) )
-///
-/// This implementation is linear in number of inputs in terms of field
-/// operations. It also has a quadratic term in primitive operations which is
-/// negligible compared to field operations.
-/// TODO: The quadratic term can be removed by precomputing the lagrange
-/// coefficients.
-pub fn interpolate_uni_poly<F: Field>(p_i: &[F], eval_at: F) -> F {
-    let start = entered_span!("sum check interpolate uni poly opt");
-
-    let len = p_i.len();
-    let mut evals = vec![];
-    let mut prod = eval_at;
-    evals.push(eval_at);
-
-    // `prod = \prod_{j} (eval_at - j)`
-    for e in 1..len {
-        let tmp = eval_at - F::from_u64(e as u64);
-        evals.push(tmp);
-        prod *= tmp;
-    }
-    let mut res = F::ZERO;
-    // we want to compute \prod (j!=i) (i-j) for a given i
-    //
-    // we start from the last step, which is
-    //  denom[len-1] = (len-1) * (len-2) *... * 2 * 1
-    // the step before that is
-    //  denom[len-2] = (len-2) * (len-3) * ... * 2 * 1 * -1
-    // and the step before that is
-    //  denom[len-3] = (len-3) * (len-4) * ... * 2 * 1 * -1 * -2
-    //
-    // i.e., for any i, the one before this will be derived from
-    //  denom[i-1] = denom[i] * (len-i) / i
-    //
-    // that is, we only need to store
-    // - the last denom for i = len-1, and
-    // - the ratio between current step and fhe last step, which is the product of (len-i) / i from
-    //   all previous steps and we store this product as a fraction number to reduce field
-    //   divisions.
-
-    let mut denom_up = field_factorial::<F>(len - 1);
-    let mut denom_down = F::ONE;
-
-    for i in (0..len).rev() {
-        res += p_i[i] * prod * denom_down * (denom_up * evals[i]).inverse();
-
-        // compute denom for the next step is current_denom * (len-i)/i
-        if i != 0 {
-            denom_up *= -F::from_u64((len - i) as u64);
-            denom_down *= F::from_u64(i as u64);
-        }
-    }
-    exit_span!(start);
-    res
+/// this implementation uses exactly 3 field inversions, which is optimal for barycentric
+///   evaluation at degree 1:
+///   - Two inverses for the distances `d0 = x - x0`, `d1 = x - x1`
+///   - One inverse for the final normalization term
+fn extrapolate_uni_poly_deg_1<F: Field>(p_i: &[F; 2], eval_at: F) -> F {
+    let x0 = F::ZERO;
+    let x1 = F::ONE;
+
+    let w0 = F::ONE;
+    let w1 = -F::ONE;
+
+    let d0 = eval_at - x0; // = eval_at
+    let d1 = eval_at - x1; // = eval_at - 1
+
+    let inv_d0 = d0.inverse();
+    let inv_d1 = d1.inverse();
```

Comment:
what `d0` or `d1` is zero?

Reply:
1. the probability of d0 or d1 to be zero is only 2/|E| which is negligible
2. it will panic on rust version verifier https://github.com/scroll-tech/Plonky3/blob/6db45dd9da0e4ef496f98c96f780a4eaf0bccd0f/field/src/field.rs#L483 but we should deal with error properly in dsl

Reply:
Since `eval_at` is the outcome of poseidon2 permutation, it's possible that hacker can take advantage of this to get the verifier (either in rust or circuit) panic. This will hurt our liveness guarantee.

Reply:
> Since `eval_at` is the outcome of poseidon2 permutation, it's possible that hacker can take advantage of this to get the verifier (either in rust or circuit) panic. This will hurt our liveness guarantee.

Do you mean $0 = \text{Poseidon}(\text{pre-image})$ ? this account for preimage-resistant cryptographic permutation and Poseidon should got the same property which make this operation infeasible